tests/Microsoft.Identity.Test.Unit/BrokerTests/WamTests.cs

```diff
@@ -1,698 +1,755 @@
+using Microsoft.Identity.Client.OAuth2;
+using Microsoft.Identity.Client.Platforms.Features.WamBroker;
+using Microsoft.Identity.Client.UI;
+using Microsoft.Identity.Client.Utils;
+using Microsoft.Identity.Test.Common.Core.Helpers;
+using Microsoft.Identity.Test.Common.Core.Mocks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using NSubstitute;
+using Windows.Security.Authentication.Web.Core;
+using Windows.Security.Credentials;
+
+namespace Microsoft.Identity.Test.Unit.BrokerTests
+{
+    [TestClass]
+    [TestCategory("Broker")]
+    public class WamTests : TestBase
+    {
+        private CoreUIParent _coreUIParent;
+        private ICoreLogger _logger;
+        private IWamPlugin _aadPlugin;
+        private IWamPlugin _msaPlugin;
+        private IWamProxy _wamProxy;
+        private IWebAccountProviderFactory _webAccountProviderFactory;
+        private IAccountPickerFactory _accountPickerFactory;
+        private WamBroker _wamBroker;
+        private SynchronizationContext _synchronizationContext;
+
+        private MsalTokenResponse _msalTokenResponse = new MsalTokenResponse
+        {
+            IdToken = MockHelpers.CreateIdToken(TestConstants.UniqueId, TestConstants.DisplayableId),
+            AccessToken = "access-token",
+            ClientInfo = MockHelpers.CreateClientInfo(),
+            ExpiresIn = 3599,
+            CorrelationId = "correlation-id",
+            RefreshToken = null, // brokers don't return RT
+            Scope = TestConstants.s_scope.AsSingleString(),
+            TokenType = "Bearer",
+            WamAccountId = "wam_account_id",
+        };
+
+        [TestInitialize]
+        public void Init()
+        {
+            _synchronizationContext = new DedicatedThreadSynchronisationContext();
+
+            _coreUIParent = new CoreUIParent() { SynchronizationContext = _synchronizationContext };
+
+            _logger = Substitute.For<ICoreLogger>();
+            _aadPlugin = Substitute.For<IWamPlugin>();
+            _msaPlugin = Substitute.For<IWamPlugin>();
+            _wamProxy = Substitute.For<IWamProxy>();
+            _webAccountProviderFactory = Substitute.For<IWebAccountProviderFactory>();
+            _accountPickerFactory = Substitute.For<IAccountPickerFactory>();
+
+            _wamBroker = new WamBroker(
+                _coreUIParent,
+                _logger,
+                _aadPlugin,
+                _msaPlugin,
+                _wamProxy,
+                _webAccountProviderFactory,
+                _accountPickerFactory);
+        }
+
+        [TestMethod]
+        public void HandleInstallUrl_Throws()
+        {
+            AssertException.Throws<NotImplementedException>(() => _wamBroker.HandleInstallUrl("http://app"));
+        }
+
+        [TestMethod]
+        [Description("For AcquireTokenSilent, plugin selection occurs based on the authority only, as this is always tenanted.")]
+        public async Task Ats_PluginSelection_Async()
```

Comment:
so if the dev passes in /common will you pull the tenantId from elsewhere?

Reply:
That is a very good question and probably one of the more complex areas of this integration. The code is [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/83acb1d345757b1df611a135105766b31c8892d6/src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/WamBroker.cs#L630) 

In particular for AcquireTokenSilent we can use: input authority and account home tenant id

if input authority is `common`, and an account was passed in, then see if home tenant ID is MSA_tenantID. However, account==null, then MSAL will log in the current user, and we have a way of figuring out if the "current user" is MSA or not - [2 liner](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/83acb1d345757b1df611a135105766b31c8892d6/src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/WebAccountProviderFactory.cs#L28)

Reply:
i see. thanks. appreciate the links